### PR TITLE
fix(evpn-bridge): fix system behaviour for pending objects

### PR DESCRIPTION
**Pull request by mardim91**
This is a fix related to the issue https://github.com/opiproject/opi-evpn-bridge/issues/388

### CODE REVIEW

File: pkg/infradb/taskmanager/taskmanager.go [hunk 1]
```diff
@@ -143,11 +161,17 @@ func (t *TaskManager) processTasks() {
...
-					log.Printf("processTasks(): No task status has been received in the channel from subscriber %+v. The task %+v will be requeued immediately Task Status %+v\n", sub, task, taskStatus)
+					log.Printf("processTasks(): No task status has been received in the channel from subscriber %+v. The task %+v will be requeued. Task Status %+v\n", sub, task, taskStatus)
+					// We keep this subIndex in order to know from which subscriber to start iterating after the requeue of the Task
+					// so we do start again from the subscriber that returned an error or was unavailable for any reason.
 					task.subIndex += i
-					go t.taskQueue.Enqueue(task)
+					task.systemTimer *= 2
+					log.Printf("processTasks(): The Task will be requeued after %+v\n", task.systemTimer)
+					time.AfterFunc(task.systemTimer, func() {
+						t.taskQueue.Enqueue(task)
```
Comment (artek-koltun) [inline pkg/infradb/taskmanager/taskmanager.go:173]: 1. if we do not receive an answer within reasonable time, does it mean that we have a problem and probably no one is handling our requests? and the best we can do is report an error and gracefully shutdown?
 
2. If we block for 30 seconds in this thread, can we handle other task responses? If we are blocked for 30 seconds, doesn't it mean that what you actually want is to wait for a response in another goroutine which could continue handling when the response is received?

File: pkg/infradb/taskmanager/taskmanager.go [hunk 2]
```diff
@@ -159,19 +183,27 @@ func (t *TaskManager) processTasks() {
...
+				// so we do start again from the subscriber that returned an error or was unavailable for any reason.
 				task.subIndex += i
-				task.retryTimer = taskStatus.component.Timer
-				log.Printf("processTasks(): The Task will be requeued after %+v\n", task.retryTimer)
-				time.AfterFunc(task.retryTimer, func() {
+				time.AfterFunc(taskStatus.component.Timer, func() {
 					t.taskQueue.Enqueue(task)
 				})
 				break loopTwo
+			default:
```
Comment (artek-koltun) [inline pkg/infradb/taskmanager/taskmanager.go:203]: as I understand if component cannot handle a task at all, it won't send response, but we still wait for 30 seconds?

**Review by artek-koltun — COMMENTED** [review]
I am wondering if we could create some unit tests which could prove that this approach work and free of race conditions. Go tests can be run with `-race` flag to catch any race conditions:
https://go.dev/doc/articles/race_detector

File: pkg/infradb/subscriberframework/eventbus/eventbus.go [hunk 3]
```diff
@@ -128,10 +129,20 @@ func (e *EventBus) subscriberExist(eventType string, moduleName string) bool {
 }
 
 // Publish api notifies the subscribers with certain eventType
-func (e *EventBus) Publish(objectData *ObjectData, subscriber *Subscriber) {
+func (e *EventBus) Publish(objectData *ObjectData, subscriber *Subscriber) error {
 	e.publishL.RLock()
 	defer e.publishL.RUnlock()
-	subscriber.Ch <- objectData
+	var err error
+	// We need the default case here as if the subscriber is busy then we will not be able to sent to the subscriber channel
```
[the remark this answers is not in the document]
Comment (mardim91) [reply pkg/infradb/subscriberframework/eventbus/eventbus.go:136]: ok Yes I can remove the comment

File: pkg/infradb/subscriberframework/eventbus/eventbus.go [hunk 4]
```diff
@@ -128,10 +129,20 @@ func (e *EventBus) subscriberExist(eventType string, moduleName string) bool {
 }
 
 // Publish api notifies the subscribers with certain eventType
-func (e *EventBus) Publish(objectData *ObjectData, subscriber *Subscriber) {
+func (e *EventBus) Publish(objectData *ObjectData, subscriber *Subscriber) error {
 	e.publishL.RLock()
```
[the remark this answers is not in the document]
Comment (mardim91) [reply pkg/infradb/subscriberframework/eventbus/eventbus.go:133]: I think you are right. The question though is if we actually need a Lock.

In a case where two goroutines try to write on the same channel I think golang is smart enough to block the second goroutine until the first one has finished writing. In that case we do not need any Lock. is that correct ?

WDYT ?

File: pkg/infradb/taskmanager/taskmanager.go [hunk 5]
```diff
@@ -123,7 +130,18 @@ func (t *TaskManager) processTasks() {
 				// (e.g. Maybe you have a timeout on the subscribers and you got the notification after the timeout have passed)
 				NotificationID: uuid.NewString(),
 			}
-			eventbus.EBus.Publish(objectData, sub)
+			if err := eventbus.EBus.Publish(objectData, sub); err != nil {
+				log.Printf("processTasks(): Notification not sent to subscriber %+v with data %+v. Subscriber is busy. The Task %+v will be requeued.\n", sub, objectData, task)
+				// We keep this subIndex in order to know from which subscriber to start iterating after the requeue of the Task
+				// so we do start again from the subscriber that returned an error or was unavailable for any reason.
+				task.subIndex += i
```
[the remark this answers is not in the document]
Comment (mardim91) [reply pkg/infradb/taskmanager/taskmanager.go:137]: No this will not happen as the next time we will take a sublist of the subscribers based on the subIndex. Then we will iterate on that sublist and that means that the i will start from 0 again.

check here: https://github.com/opiproject/opi-evpn-bridge/blob/main/pkg/infradb/taskmanager/taskmanager.go#L114

File: pkg/infradb/taskmanager/taskmanager.go [hunk 6]
```diff
@@ -123,7 +130,18 @@ func (t *TaskManager) processTasks() {
...
+				task.systemTimer *= 2
```
[the remark this answers is not in the document]
Comment (mardim91) [reply pkg/infradb/taskmanager/taskmanager.go:138]: - I can change the name of the var that is no problem.
- The timer will increase every time that we try to publish but the channel is busy and we need to reque the task. So it is working this way:

1. We publish
2. An error is returned because the channel is busy
3. We increase the timer
4. We wait for the timer to expire in a goroutine and then we requeue the task

File: pkg/infradb/taskmanager/taskmanager.go (hunk 2 above)
Comment (mardim91) [reply pkg/infradb/taskmanager/taskmanager.go:203]: This code here is for the case where the component has handled the task and has sent a Status back but this is neither Success or Error (I do not expect something like this to happen but still I have added this default case for this unlikely scenario). In this case we just drop the Task and move on to the next Task in the queue

**Comment by mardim91** [comment]
> I am wondering if we could create some unit tests which could prove that this approach work and free of race conditions. Go tests can be run with `-race` flag to catch any race conditions: https://go.dev/doc/articles/race_detector

I have tested this code locally in order to check if the different corner cases that are resolved by this code here are working properly. Everything was working as it should be. 

The unit tests would be a very usefull tool to add in order not to test all the cases by hand every time but right now I do not have time to work on that so I will propose to open an issue so we can address into the future when I find some time if that is allright.

File: pkg/infradb/subscriberframework/eventbus/eventbus.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (mardim91) [reply pkg/infradb/subscriberframework/eventbus/eventbus.go:136]: I would like though to keep the comment in the task manager as it describes a specific corner case and can help the reader understand better what this is all about.

File: pkg/infradb/taskmanager/taskmanager.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (artek-koltun) [reply pkg/infradb/taskmanager/taskmanager.go:139]: The flow you described is clear. I want to know if we ever stop increasing it. Will it make to wait for hours? days? years?

File: pkg/infradb/taskmanager/taskmanager.go (hunk 1 above)
Comment (artek-koltun) [reply pkg/infradb/taskmanager/taskmanager.go:174]: > Now if the subscriber is completely dead and cannot send a response back at all we need to put in plase some resiliency mechanisms to handle this case btu this is something to investigate and implement in the future currently we do not look into that corner case.

Issue is needed

>  The system is designed this way where we send a task to a subscriber and then we just wait for that subsriber to respond before we move on to the next subscriber.

Sounds like a sequential flow. Do we need channels then? Pls consider. It would make all the things much simpler

File: pkg/infradb/taskmanager/taskmanager.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (artek-koltun) [reply pkg/infradb/taskmanager/taskmanager.go:138]: Sounds a bit error-prone
What if we, instead of index calculations, copy the rest of subscribers into a task, so it could continue where it stopped?
Any other means?

[the remark this answers is not in the document]
Comment (mardim91) [reply pkg/infradb/taskmanager/taskmanager.go:138]: To be honest I like the solution as it is today. It looks cleaner to me. It is tested and it works so nice so I would not like to change it if that is ok with you. I can put a comment if you want so people can understand what this index calculation is all about. I do not like  so much to be honest to keep sublists all the time for the remaining subscribers to me that is a bit more error prone. Maybe we can revisit this issue in the future as this bug fix here is not related to the subIndex. The SubIndex was allready there from the beggining

File: pkg/infradb/taskmanager/taskmanager.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (mardim91) [reply pkg/infradb/taskmanager/taskmanager.go:139]: Currently we never stop increasing it. But this is a known issue that we have just not addressed do far. We are planning to address this in the future. I can open an issue to track it.

File: pkg/infradb/taskmanager/taskmanager.go (hunk 1 above)
Comment (mardim91) [reply pkg/infradb/taskmanager/taskmanager.go:174]: I think channels and notifications make the whole system scale better. The implementation that we have so far it just uses one task queue and one process that drains the queue. Maybe in the future if we hit any limitation we could use more processes to drain the queue of the tasks and I think this will scale easier when we are using channels and notifications. So I would prefer keeping it this way.

**Comment by mardim91** [comment]
> > > I am wondering if we could create some unit tests which could prove that this approach work and free of race conditions. Go tests can be run with `-race` flag to catch any race conditions: https://go.dev/doc/articles/race_detector
> > 
> > 
> > I have tested this code locally in order to check if the different corner cases that are resolved by this code here are working properly. Everything was working as it should be.
> > The unit tests would be a very usefull tool to add in order not to test all the cases by hand every time but right now I do not have time to work on that so I will propose to open an issue so we can address into the future when I find some time if that is allright.
> 
> if we leave it for an issue, it can stay there for a very long time, especially if no one have an incline for unit testing. Maybe someone could help you with that so we do not defer that?

ok let me see what I can do about this

File: pkg/infradb/subscriberframework/eventbus/eventbus.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (mardim91) [reply pkg/infradb/subscriberframework/eventbus/eventbus.go]: Yes I think the Lock is not needed here we can remove it

[the remark this answers is not in the document]
Comment (artek-koltun) [reply pkg/infradb/subscriberframework/eventbus/eventbus.go]: pls remove then

File: pkg/infradb/taskmanager/taskmanager.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (artek-koltun) [reply pkg/infradb/taskmanager/taskmanager.go:138]: For me it is not. I already misinterpreted what is happening here.

> I can put a comment if you want so people can understand what this index calculation is all about.

Please do. I am also wondering if it could be a single place like a function to "re-queue task"

> Maybe we can revisit this issue in the future as this bug fix here is not related to the subIndex. The SubIndex was allready there from the beggining

Please create an issue

File: pkg/infradb/taskmanager/taskmanager.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (artek-koltun) [reply pkg/infradb/taskmanager/taskmanager.go:139]: Pls create an issue

File: pkg/infradb/taskmanager/taskmanager.go (hunk 1 above)
Comment (mardim91) [reply pkg/infradb/taskmanager/taskmanager.go:174]: Maybe I didn't get your question correct at the beggining. I apologize for this

The channels are not so much related to the sequential flow as to keep the core system as much as was possible agnostic to the plugins that we need to send and receive notifications. The plugins are responsible to configuring the underline system of their responsibility (e.g. FRRModule for FRR, LinuxModule for Linux etc...). The core system doesn't need to know what are those plugins or what they exactly do but the only thing that they need to know is towards which subscribers they need to send notifications. The channels help in this agnostic notion as well as to the communication between the different go routines as each subscriber runs as a different go routine. 

Also the sequential flow it is there because the different plugins have some sort of dependency into eachother. That dependency is expressed by the sequential flow and that is why we want the first plugin to succeed before we move to the second one. Because for instance if the Linux Vendor Module wants to attach an interface into a bridge which bridge has been created before by General Linux Module if the sequential flow is not there the Linux Vendor module call will fail as it has a dependency to the General Linux module operation first to create the bridge. 

This is a significant design choice which has been presented to the TSC before the implementation and we have decided that was resonable to move forward with it . Also we think that is a good design and works well so far and I do not really agree that the channels make the system complicated as they serve the architecture well.

Comment (artek-koltun) [reply pkg/infradb/taskmanager/taskmanager.go:174]: > The channels are not so much related to the sequential flow as to keep the core system as much as was possible agnostic to the plugins that we need to send and receive notifications. 

The same level of agnosticism can be achieved with a function calling. The channels do not add more agnosticism here.
By the [definition](https://gobyexample.com/channels): Channels are the pipes that connect concurrent goroutines. You can send values into channels from one goroutine and receive those values into another goroutine.

> as well as to the communication between the different go routines as each subscriber runs as a different go routine.

They can be run in different routines, even if we call them by a function (you can use mutex to sync or send to channel or execute right in the same thread context)

> Also the sequential flow it is there because the different plugins have some sort of dependency into eachother. That dependency is expressed by the sequential flow and that is why we want the first plugin to succeed before we move to the second one.

We send a msg, and go to sleep waiting for a plugin to complete its job transforming our flow into sequential

> Because for instance if the Linux Vendor Module wants to attach an interface into a bridge which bridge has been created before by General Linux Module if the sequential flow is not there the Linux Vendor module call will fail as it has a dependency to the General Linux module operation first to create the bridge.

What if we do not register General Linux Module at all so it won't receive notifications at all?

> Also we think that is a good design and works well so far and I do not really agree that the channels make the system complicated as they serve the architecture well.

It might be. But from the chunk I see, it looks like you need a sequential flow and channels add complexity

**Comment by mardim91** [comment]
> > > > I am wondering if we could create some unit tests which could prove that this approach work and free of race conditions. Go tests can be run with `-race` flag to catch any race conditions: https://go.dev/doc/articles/race_detector
> > > 
> > > 
> > > I have tested this code locally in order to check if the different corner cases that are resolved by this code here are working properly. Everything was working as it should be.
> > > The unit tests would be a very usefull tool to add in order not to test all the cases by hand every time but right now I do not have time to work on that so I will propose to open an issue so we can address into the future when I find some time if that is allright.
> > 
> > 
> > if we leave it for an issue, it can stay there for a very long time, especially if no one have an incline for unit testing. Maybe someone could help you with that so we do not defer that?
> 
> ok let me see what I can do about this

Hello @artek-koltun 

I have one question regarding race detection. From what I understand from the link that you have provided I should not create any unit tests for catching race conditions but the only thing that I should do is to create some automatic github jobs that will build/run the programm with the --race parameter on (go build --race/ go run --race) is that correct or ? Can you please share your opinion on this ?

**Comment by artek-koltun** [comment]
> > > > > I am wondering if we could create some unit tests which could prove that this approach work and free of race conditions. Go tests can be run with `-race` flag to catch any race conditions: https://go.dev/doc/articles/race_detector
> > > > 
> > > > 
> > > > I have tested this code locally in order to check if the different corner cases that are resolved by this code here are working properly. Everything was working as it should be.
> > > > The unit tests would be a very usefull tool to add in order not to test all the cases by hand every time but right now I do not have time to work on that so I will propose to open an issue so we can address into the future when I find some time if that is allright.
> > > 
> > > 
> > > if we leave it for an issue, it can stay there for a very long time, especially if no one have an incline for unit testing. Maybe someone could help you with that so we do not defer that?
> > 
> > 
> > ok let me see what I can do about this
> 
> Hello @artek-koltun
> 
> I have one question regarding race detection. From what I understand from the link that you have provided I should not create any unit tests for catching race conditions but the only thing that I should do is to create some automatic github jobs that will build/run the programm with the --race parameter on (go build --race/ go run --race) is that correct or ? Can you please share your opinion on this ?

We already run our uts with that flag
https://github.com/opiproject/actions/blob/ae6a52f2fcf782b0c29df17e3c61475096ebb5d2/.github/workflows/go.yml#L35-L37

However, take in mind that it is not a compile time tool. It is a runtime tool. We need to execute a chunk of code, which can have race issues. Thus, we need a dedicated unit test which covers the code with the issue.

**Comment by mardim91** [comment]
> > > > > > I am wondering if we could create some unit tests which could prove that this approach work and free of race conditions. Go tests can be run with `-race` flag to catch any race conditions: https://go.dev/doc/articles/race_detector
> > > > > 
> > > > > 
> > > > > I have tested this code locally in order to check if the different corner cases that are resolved by this code here are working properly. Everything was working as it should be.
> > > > > The unit tests would be a very usefull tool to add in order not to test all the cases by hand every time but right now I do not have time to work on that so I will propose to open an issue so we can address into the future when I find some time if that is allright.
> > > > 
> > > > 
> > > > if we leave it for an issue, it can stay there for a very long time, especially if no one have an incline for unit testing. Maybe someone could help you with that so we do not defer that?
> > > 
> > > 
> > > ok let me see what I can do about this
> > 
> > 
> > Hello @artek-koltun
> > I have one question regarding race detection. From what I understand from the link that you have provided I should not create any unit tests for catching race conditions but the only thing that I should do is to create some automatic github jobs that will build/run the programm with the --race parameter on (go build --race/ go run --race) is that correct or ? Can you please share your opinion on this ?
> 
> We already run our uts with that flag https://github.com/opiproject/actions/blob/ae6a52f2fcf782b0c29df17e3c61475096ebb5d2/.github/workflows/go.yml#L35-L37
> 
> However, take in mind that it is not a compile time tool. It is a runtime tool. We need to execute a chunk of code, which can have race issues. Thus, we need a dedicated unit test which covers the code with the issue.

Hello @artek-koltun,

ok so from what I understand the unit tests are independent of the --race condition. The --race condition checks for races in specific parts of the code which code the unit tests are checking. Right ?

Also we have an agreeement to create unit tests no worries. We will create unit tests for the units (functions) that we are touching in the PR if that is allright

**Comment by artek-koltun** [comment]
> Hello @artek-koltun,
> 
> ok so from what I understand the unit tests are independent of the --race condition. The --race condition checks for races in specific parts of the code which code the unit tests are checking. Right ?
> 
> Also we have an agreeement to create unit tests no worries. We will create unit tests for the units (functions) that we are touching in the PR if that is allright

Hej,
You can use the following algorithm to write your unit test:
1. Write a test which would fail with -race flag without a fix. (you might need to run multiple components/goroutines to create an env provoking the error)
2. Write a fix
3. Commit and push.

**Comment by artek-koltun** [comment]
> > > Hello @artek-koltun,
> > > ok so from what I understand the unit tests are independent of the --race condition. The --race condition checks for races in specific parts of the code which code the unit tests are checking. Right ?
> > > Also we have an agreeement to create unit tests no worries. We will create unit tests for the units (functions) that we are touching in the PR if that is allright
> > 
> > 
> > Hej, You can use the following algorithm to write your unit test:
> > 
> > 1. Write a test which would fail with -race flag without a fix. (you might need to run multiple components/goroutines to create an env provoking the error)
> > 2. Write a fix
> > 3. Commit and push.
> 
> Hello @artek-koltun ,
> 
> What do you mean write a test that would fail with -race flag. I do not think that we have any race problems in the code today so we can catch them by unit test and fix them afterwards.
> 
> What we are planning to so is that we will write unit tests for the code changes in this commit, we will run the tests with -race flag on and if in the meanwhile find any race problem then we will push a fix.
> 
> I hope we are in the same page.

I might have lost the context. Pls create ut to the behavior you are fixing.

File: pkg/infradb/taskmanager/taskmanager.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (mardim91) [reply pkg/infradb/taskmanager/taskmanager.go:143]: Done

https://github.com/opiproject/opi-evpn-bridge/issues/393

File: pkg/infradb/taskmanager/taskmanager.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (mardim91) [reply pkg/infradb/taskmanager/taskmanager.go:144]: https://github.com/opiproject/opi-evpn-bridge/issues/394

File: pkg/infradb/subscriberframework/eventbus/event_bus_test.go [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+	select {
+	case _, ok := <-subscriber.Ch:
+		if ok {
+			t.Errorf("Subscriber's channel should be closed, but it's not")
+		}
+	default:
+	}
+}
+
+func TestSubscriberExist(t *testing.T) {
```
Comment (artek-koltun) [inline pkg/infradb/subscriberframework/eventbus/event_bus_test.go:163]: AlreadyExist

File: pkg/infradb/taskmanager/taskmanager_test.go [hunk 8]
```diff
@@ -0,0 +1,295 @@
...
+	case "testEventTimeout":
+		handletestEventTimeout()
+	case "testEventNotificationIdNotMatching":
+		handletestNotificationIDNotMatching(objectData)
+	default:
+		log.Printf("LCI: error: Unknown event type %s", eventType)
+	}
+}
+
+func TestTaskManager_CreateTask(t *testing.T) {
```
Comment (artek-koltun) [inline pkg/infradb/taskmanager/taskmanager_test.go:119]: Please use the same naming for all uts

File: pkg/infradb/subscriberframework/eventbus/event_bus_test.go [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+	receivedEvents []*ObjectData
+	sync.Mutex
+}
+
+func (h *ModulelciHandler) HandleEvent(eventType string, objectData *ObjectData) {
+	h.Lock()
+	defer h.Unlock()
+	h.receivedEvents = append(h.receivedEvents, objectData)
+	switch eventType {
+	case "testEvent":
```
Comment (artek-koltun) [inline pkg/infradb/subscriberframework/eventbus/event_bus_test.go:20]: why not create constants for those values? It would be much harder to make a mistake by a typo

File: pkg/infradb/taskmanager/taskmanager_test.go [hunk 10]
```diff
@@ -0,0 +1,295 @@
...
+	subscriber := &eventbus.Subscriber{
+		Name:     "testSubscriber",
+		Ch:       make(chan interface{}),
+		Quit:     make(chan bool),
+		Priority: 1,
+	}
+	subs := []*eventbus.Subscriber{subscriber}
+
+	TaskMan.StartTaskManager()
```
Comment (artek-koltun) [inline pkg/infradb/taskmanager/taskmanager_test.go:128]: it is better to use its own instance of a manager per test, then a global one

File: pkg/infradb/taskmanager/taskmanager_test.go [hunk 11]
```diff
@@ -0,0 +1,295 @@
...
+	TaskMan.CreateTask("testTask", "testType", "testVersion", subs)
+
+	time.Sleep(100 * time.Millisecond)
+
+	task := TaskMan.taskQueue.Dequeue()
```
Comment (artek-koltun) [inline pkg/infradb/taskmanager/taskmanager_test.go:133]: why we need to touch the internals instead of waiting when TaskMan sends it to us?

File: pkg/infradb/taskmanager/taskmanager_test.go [hunk 12]
```diff
@@ -0,0 +1,295 @@
...
+	assert.Equal(t, "testVersion", task.resourceVersion)
+	assert.Equal(t, subs, task.subs)
+}
+
+func TestTaskManager_CompSuccess(t *testing.T) {
+	var wg sync.WaitGroup
+
+	TaskMan.StartTaskManager()
+	handler := &ModulelciHandler{}
+	wg.Add(1)
```
Comment (artek-koltun) [inline pkg/infradb/taskmanager/taskmanager_test.go:146]: maybe you want to have this wg within StartSubscriber, so it exits only when we are listening for event?

File: pkg/infradb/taskmanager/taskmanager_test.go [hunk 13]
```diff
@@ -0,0 +1,295 @@
...
+	subscribers := eventbus.EBus.GetSubscribers("testEventCompSuccess")
+	if len(subscribers) == 0 {
+		t.Fatalf("No subscribers found for event type 'testEvent'")
+	}
+	TaskMan.CreateTask("testTaskCompSuccess", "testTypeCompSuccess", "testVersionCompSuccess", subscribers)
+
+	select {
+	case task := <-TaskMan.taskQueue.channel:
+		if task.name == "testTaskCompSuccess" {
+			t.Errorf("assert failed:")
```
Comment (artek-koltun) [inline pkg/infradb/taskmanager/taskmanager_test.go:163]: what exactly happened here?

File: pkg/infradb/taskmanager/taskmanager_test.go [hunk 14]
```diff
@@ -0,0 +1,295 @@
...
+	var wg sync.WaitGroup
+
+	wg.Add(1)
+
+	go handleTestEventbusy(&wg)
+
+	wg.Wait()
+
+	retValMu.Lock()
+	actualRetVal := retVal
```
Comment (artek-koltun) [inline pkg/infradb/taskmanager/taskmanager_test.go:291]: atomic?

File: pkg/infradb/subscriberframework/eventbus/event_bus_test.go [hunk 15]
```diff
@@ -0,0 +1,181 @@
...
+	EventBus.StartSubscriber("testModule2", "testEventpriority", 1, handler2)
+
+	time.Sleep(100 * time.Millisecond)
+
+	subscribers := EventBus.GetSubscribers("testEventpriority")
+	if len(subscribers) != 2 {
+		t.Errorf("Expected 2 subscribers, got %d", len(subscribers))
+	}
+	if subscribers[0].Priority > subscribers[1].Priority {
+		t.Errorf("Subscribers are not sorted by priority")
```
[the remark this answers is not in the document]
Comment (atulpatel261194) [reply pkg/infradb/subscriberframework/eventbus/event_bus_test.go:91]: they are sorted each time.

File: pkg/infradb/taskmanager/taskmanager_test.go (hunk 14 above)
Comment (atulpatel261194) [reply pkg/infradb/taskmanager/taskmanager_test.go:285]: i believe we can keep lock as its limited to this test only.

File: pkg/infradb/taskmanager/taskmanager_test.go (hunk 13 above)
Comment (atulpatel261194) [reply pkg/infradb/taskmanager/taskmanager_test.go:163]: task dequeued, processed by comp and queue become empty as it was successful.

File: pkg/infradb/taskmanager/taskmanager_test.go (hunk 12 above)
Comment (atulpatel261194) [reply pkg/infradb/taskmanager/taskmanager_test.go:146]: then we have to change the whole design. which will impact all the other modules/pkgs also ?

File: pkg/infradb/taskmanager/taskmanager_test.go (hunk 11 above)
Comment (atulpatel261194) [reply pkg/infradb/taskmanager/taskmanager_test.go]: here we are just checking for this case if task was created or not.

File: pkg/infradb/taskmanager/taskmanager_test.go (hunk 8 above)
Comment (atulpatel261194) [reply pkg/infradb/taskmanager/taskmanager_test.go]: Done

File: pkg/infradb/subscriberframework/eventbus/event_bus_test.go (hunk 7 above)
Comment (atulpatel261194) [reply pkg/infradb/subscriberframework/eventbus/event_bus_test.go]: Done

File: pkg/infradb/subscriberframework/eventbus/event_bus_test.go [hunk 16]
```diff
@@ -0,0 +1,181 @@
...
+		t.Errorf("No subscribers found for event type 'testEvent'")
+	}
+	subscriber := subscribers[0]
+
+	err := EventBus.Publish(objectData, subscriber)
+	if err != nil {
+		t.Errorf("Publish() failed with error: %v", err)
+	}
+
+	time.Sleep(100 * time.Millisecond)
```
[the remark this answers is not in the document]
Comment (atulpatel261194) [reply pkg/infradb/subscriberframework/eventbus/event_bus_test.go]: by doing that we are kind of diverting the whole test. then it will never fail if there is any change in task manager code.

File: pkg/infradb/subscriberframework/eventbus/event_bus_test.go [hunk 17]
```diff
@@ -0,0 +1,181 @@
...
+	"testing"
+	"time"
+)
+
+type ModulelciHandler struct {
+	receivedEvents []*ObjectData
+	sync.Mutex
+}
+
+func (h *ModulelciHandler) HandleEvent(eventType string, objectData *ObjectData) {
```
[the remark this answers is not in the document]
Comment (atulpatel261194) [reply pkg/infradb/subscriberframework/eventbus/event_bus_test.go]: done

File: pkg/infradb/subscriberframework/eventbus/event_bus_test.go [hunk 18]
```diff
@@ -0,0 +1,181 @@
...
+	subscribers := EventBus.GetSubscribers("testEventpriority")
+	if len(subscribers) != 2 {
+		t.Errorf("Expected 2 subscribers, got %d", len(subscribers))
+	}
+	if subscribers[0].Priority > subscribers[1].Priority {
+		t.Errorf("Subscribers are not sorted by priority")
+	}
+
+	for _, sub := range subscribers {
+		EventBus.Unsubscribe(sub)
```
[the remark this answers is not in the document]
Comment (atulpatel261194) [reply pkg/infradb/subscriberframework/eventbus/event_bus_test.go:95]: Defer wont help as all test have different leftovers

File: pkg/infradb/subscriberframework/eventbus/event_bus_test.go [hunk 19]
```diff
@@ -0,0 +1,181 @@
...
+	subscriber := subscribers[0]
+
+	err := EventBus.Publish(objectData, subscriber)
+	if err != nil {
+		t.Errorf("Publish() failed with error: %v", err)
+	}
+
+	time.Sleep(100 * time.Millisecond)
+	handler.Lock()
+	if len(handler.receivedEvents) != 1 || handler.receivedEvents[0] != objectData {
```
[the remark this answers is not in the document]
Comment (atulpatel261194) [reply pkg/infradb/subscriberframework/eventbus/event_bus_test.go]: Done

File: pkg/infradb/subscriberframework/eventbus/event_bus_test.go [hunk 20]
```diff
@@ -0,0 +1,181 @@
...
+	defer h.Unlock()
+	h.receivedEvents = append(h.receivedEvents, objectData)
+	switch eventType {
+	case "testEvent":
+	case "testEventpriority":
+	case "testEventChBusy":
+	case "testEventUnsub":
+		log.Printf("received event type %s", eventType)
+	default:
+		log.Printf("error: Unknown event type %s", eventType)
```
[the remark this answers is not in the document]
Comment (atulpatel261194) [reply pkg/infradb/subscriberframework/eventbus/event_bus_test.go]: Done

File: pkg/infradb/subscriberframework/eventbus/event_bus_test.go [hunk 21]
```diff
@@ -0,0 +1,181 @@
...
+	}
+}
+
+func TestSubscribeAndPublish(t *testing.T) {
+	handler := &ModulelciHandler{}
+
+	EventBus := NewEventBus()
+	EventBus.StartSubscriber("testModule", "testEvent", 1, handler)
+
+	time.Sleep(100 * time.Millisecond)
```
[the remark this answers is not in the document]
Comment (atulpatel261194) [reply pkg/infradb/subscriberframework/eventbus/event_bus_test.go]: sleep is just for allow time to get the go routine ready.

File: pkg/infradb/taskmanager/taskmanager_test.go (hunk 13 above)
Comment (artek-koltun) [reply pkg/infradb/taskmanager/taskmanager_test.go:163]: I am wondering if we could test it without going so deep into the implementation details? By calling `processTasks`?

File: pkg/infradb/taskmanager/taskmanager_test.go (hunk 14 above)
Comment (artek-koltun) [reply pkg/infradb/taskmanager/taskmanager_test.go:285]: You use mutex and workgroup to find out if your var was changed. I think it is overkill. Using a chan would be enough (you pass a chan into `handleTestEventbusy`, you wait in select for chan be not empty or timeout, `handleTestEventbusy` sends into chan when ready)

File: pkg/infradb/taskmanager/taskmanager_test.go (hunk 10 above)
Comment (artek-koltun) [reply pkg/infradb/taskmanager/taskmanager_test.go]: not addressed

File: pkg/infradb/subscriberframework/eventbus/event_bus_test.go (hunk 9 above)
Comment (artek-koltun) [reply pkg/infradb/subscriberframework/eventbus/event_bus_test.go]: line 101
```go
func (h *moduleCiHandler) HandleEvent(eventType string, objectData *eventbus.ObjectData) {
	h.receivedEvents = append(h.receivedEvents, objectData)
	switch eventType {
	case "testEvent":
		handleTestEvent(objectData)
	case "testEventCompSuccess":
		handleTestEventCompSuccess(objectData)
	case "testEventError":
		handleTestEventError(objectData)
	case "testEventTimeout":
		handletestEventTimeout()
	case "testEventNotificationIdNotMatching":
		handletestNotificationIDNotMatching(objectData)
	default:
		log.Printf("LCI: error: Unknown event type %s", eventType)
	}
}
```

File: pkg/infradb/subscriberframework/eventbus/event_bus_test.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (artek-koltun) [reply pkg/infradb/subscriberframework/eventbus/event_bus_test.go]: I do not understand...
`HandleEvent` is part of a test, and we are free to organize it as we need
The current execution time of the test is not appropriate: [159.629s](https://github.com/opiproject/opi-evpn-bridge/actions/runs/11249540919/job/31276636859?pr=391#step:5:296) and I think we could do better here instead of using sleeps

File: pkg/infradb/subscriberframework/eventbus/event_bus_test.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (artek-koltun) [reply pkg/infradb/subscriberframework/eventbus/event_bus_test.go]: Please check naming for other variables/functions. I saw some upper case function variable names

File: pkg/infradb/subscriberframework/eventbus/event_bus_test.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (artek-koltun) [reply pkg/infradb/subscriberframework/eventbus/event_bus_test.go]: What if we are not lucky, and it is ready in 1 second? Using sleeps is bad. It is not reliable, and it makes the tests slow. At this moment, the execution time of some tests is ~[159sec](https://github.com/opiproject/opi-evpn-bridge/actions/runs/11249540919/job/31276636859?pr=391#step:5:296)
Use chan or other synchronization mechanisms

**Review by artek-koltun — CHANGES_REQUESTED** [review]
Unit tests must be fast:
https://github.com/opiproject/opi-evpn-bridge/actions/runs/11249540919/job/31276636859?pr=391#step:5:296
